pkg/securitypolicy/securitypolicyenforcer_rego.go

```diff
@@ -491,21 +541,44 @@ func (policy *regoEnforcer) enforce(enforcementPoint string, input map[string]in
 		return err
 	}
 
+	removeModule := false
+	if enforcementPoint == "load_fragment" {
+		if addModule, ok := results["add_module"].(bool); ok {
+			if !addModule {
+				removeModule = true
+			}
+		} else {
+			removeModule = true
+		}
+	}
+
 	if allowed {
 		err = policy.updateMetadata(results)
 		if err != nil {
 			return fmt.Errorf("unable to update metadata: %w", err)
 		}
-		return nil
+	} else {
+		err = policy.getReasonNotAllowed(enforcementPoint, input)
+	}
+
+	if removeModule {
+		delete(policy.modules, moduleID(input["issuer"].(string), input["feed"].(string)))
+		if compileError := policy.compile(); compileError != nil {
```

Comment:
it's a bit unclear to me when the compilation should happen? For example, if we have 2 fragments with `ns1` and `ns2` namespaces accordingly. The first fragment returns `add_module: true` and the other one `add_module: false`. When loading the first fragment, we won't hit this use case and won't re-compile and I guess access the fragment code via `rego.Module(...)(query)`. After we load the second fragment, we'll hit this code path and re-compile the policy with `ns1` module presumably?

Reply:
Yeah I need to tighten up the compilation logic as part of the minor refactor I'm doing in response to your other comment.